rust/otap-dataflow/crates/engine/src/channel_metrics.rs

```diff
@@ -0,0 +1,310 @@
+    }
+
+    #[inline]
+    pub(crate) fn record_send_error_full(&mut self) {
+        self.metrics.send_error_full.inc();
+    }
+
+    #[inline]
+    pub(crate) fn record_send_error_closed(&mut self) {
+        self.metrics.send_error_closed.inc();
+    }
+
+    #[inline]
+    pub(crate) fn report(
+        &mut self,
+        metrics_reporter: &mut MetricsReporter,
+    ) -> Result<(), TelemetryError> {
+        metrics_reporter.report(&mut self.metrics)
+    }
+}
+
+pub(crate) struct ChannelReceiverMetricsState {
+    metrics: MetricSet<ChannelReceiverMetrics>,
+    capacity: u64,
+}
+
+impl ChannelReceiverMetricsState {
+    pub(crate) fn new(metrics: MetricSet<ChannelReceiverMetrics>, capacity: u64) -> Self {
+        Self { metrics, capacity }
+    }
+
+    #[inline]
+    pub(crate) fn record_recv_ok(&mut self) {
+        self.metrics.recv_count.inc();
+    }
+
+    #[inline]
+    pub(crate) fn record_recv_error_empty(&mut self) {
+        self.metrics.recv_error_empty.inc();
+    }
+
+    #[inline]
+    pub(crate) fn record_recv_error_closed(&mut self) {
+        self.metrics.recv_error_closed.inc();
+    }
+
+    #[inline]
+    pub(crate) fn report(
+        &mut self,
+        metrics_reporter: &mut MetricsReporter,
+    ) -> Result<(), TelemetryError> {
+        self.metrics.capacity.set(self.capacity);
+        metrics_reporter.report(&mut self.metrics)
+    }
+}
+
+pub(crate) type LocalChannelSenderMetricsHandle = Rc<RefCell<ChannelSenderMetricsState>>;
+pub(crate) type LocalChannelReceiverMetricsHandle = Rc<RefCell<ChannelReceiverMetricsState>>;
+pub(crate) type SharedChannelSenderMetricsHandle = Arc<Mutex<ChannelSenderMetricsState>>;
+pub(crate) type SharedChannelReceiverMetricsHandle = Arc<Mutex<ChannelReceiverMetricsState>>;
+
+#[derive(Clone)]
+pub(crate) enum ChannelMetricsHandle {
+    LocalSender(LocalChannelSenderMetricsHandle),
+    SharedSender(SharedChannelSenderMetricsHandle),
+    LocalReceiver(LocalChannelReceiverMetricsHandle),
+    SharedReceiver(SharedChannelReceiverMetricsHandle),
+}
+
+impl ChannelMetricsHandle {
+    #[inline]
+    pub(crate) fn report(
+        &self,
+        metrics_reporter: &mut MetricsReporter,
+    ) -> Result<(), TelemetryError> {
+        match self {
+            ChannelMetricsHandle::LocalSender(metrics) => match metrics.try_borrow_mut() {
+                Ok(mut metrics) => metrics.report(metrics_reporter),
+                Err(_) => Ok(()),
+            },
+            ChannelMetricsHandle::SharedSender(metrics) => match metrics.lock() {
+                Ok(mut metrics) => metrics.report(metrics_reporter),
+                Err(_) => Ok(()),
+            },
+            ChannelMetricsHandle::LocalReceiver(metrics) => match metrics.try_borrow_mut() {
+                Ok(mut metrics) => metrics.report(metrics_reporter),
+                Err(_) => Ok(()),
+            },
+            ChannelMetricsHandle::SharedReceiver(metrics) => match metrics.lock() {
```

Comment:
same comment to use try_lock instead of lock?

Reply:
Updated